File: src/utils/OllamaProvider.js
/**
 * OllamaProvider - Local AI/embedding provider using Ollama
 * 
 * Perfect for:
 * - Privacy-conscious users (data stays local)
 * - Offline usage
 * - Cost savings (no API fees)
 * 
 * Setup:
 *   1. Install Ollama: https://ollama.com
 *   2. Pull models: `ollama pull nomic-embed-text` and `ollama pull llama3.2`
 *   3. Set OLLAMA_HOST if not running on localhost
 * 
 * Environment:
 *   OLLAMA_HOST=http://localhost:11434
 *   OLLAMA_EMBEDDING_MODEL=nomic-embed-text
 *   OLLAMA_CHAT_MODEL=llama3.2
 */

import { IntelligentProvider } from '../interfaces/EmbeddingProvider.js';

export class OllamaProvider extends IntelligentProvider {
  constructor(options = {}) {
    super({
      name: 'Ollama',
      dimensions: options.dimensions || 768, // nomic-embed-text is 768d
      ...options
    });
    
    this.baseUrl = options.baseUrl || process.env.OLLAMA_HOST || 'http://localhost:11434';
    this.embeddingModel = options.embeddingModel || process.env.OLLAMA_EMBEDDING_MODEL || 'nomic-embed-text';
    this.chatModel = options.chatModel || process.env.OLLAMA_CHAT_MODEL || 'llama3.2';
  }

  isAvailable() {
    // Ollama is available if we can reach the server
    return this.checkConnection().catch(() => false);
  }

  async checkConnection() {
    try {
      const response = await fetch(`${this.baseUrl}/api/tags`, { 
        method: 'GET',
        signal: AbortSignal.timeout(5000)
      });
      return response.ok;
    } catch {
      return false;
    }
  }

  async embed(text) {
    try {
      const response = await fetch(`${this.baseUrl}/api/embeddings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.embeddingModel,
          prompt: text
        })
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Ollama embedding error: ${error}`);
      }

      const data = await response.json();
      return data.embedding;
    } catch (err) {
      console.error(`[OllamaProvider] Embedding failed: ${err.message}`);
      return null;
    }
  }

  async embedBatch(texts) {
    // Ollama doesn't support batch embedding, so we do sequential
    const results = [];
    for (const text of texts) {
      const embedding = await this.embed(text);
      results.push(embedding);
      // Small delay to avoid overwhelming local server
      await new Promise(r => setTimeout(r, 10));
    }
    return results;
  }

  async autoTag(key, content) {
    try {
      const prompt = `Generate 5-10 relevant tags for this content.
Title: ${key}
Content: ${content.substring(0, 2000)}

Return ONLY a JSON array of lowercase, hyphenated tags.
Example: ["machine-learning", "python", "tutorial"]`;

      const response = await fetch(`${this.baseUrl}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.chatModel,
          prompt,
          stream: false,
          format: 'json'
        })
      });

      if (!response.ok) return [];

      const data = await response.json();
      const parsed = JSON.parse(data.response);
      return Array.isArray(parsed) ? parsed : parsed.tags || [];
    } catch (err) {
      console.error(`[OllamaProvider] Auto-tag failed: ${err.message}`);
      return [];
    }
  }

  async synthesize(content, task, style = '') {
    const styleGuides = {
      beginner: 'Use simple language. Explain like I\'m 5.',
      detailed: 'Be thorough with examples.',
      concise: 'Be brief. Maximum 3 sentences.'
    };

    const prompt = `${styleGuides[style] || ''}

Context from notes:
${content.substring(0, 4000)}

Task: ${task}

Provide a helpful response based on the context.`;

    try {
      const response = await fetch(`${this.baseUrl}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.chatModel,
          prompt,
          stream: false
        })
      });

      if (!response.ok) {
        throw new Error(await response.text());
      }

      const data = await response.json();
      return data.response.trim();
    } catch (err) {
      console.error(`[OllamaProvider] Synthesis failed: ${err.message}`);
      return content.substring(0, 1500);
    }
  }

  async extractGraphEntities(content) {
    const prompt = `Extract key concepts and relationships from this text as JSON.

Text: ${content.substring(0, 3000)}

Return format:
{
  "concepts": ["concept1", "concept2", ...],
  "relationships": [
    {"from": "concept1", "to": "concept2", "type": "related_to", "strength": 0.8}
  ]
}`;

    try {
      const response = await fetch(`${this.baseUrl}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.chatModel,
          prompt,
          stream: false,
          format: 'json'
        })
      });

      if (!response.ok) return { concepts: [], relationships: [] };

      const data = await response.json();
      const parsed = JSON.parse(data.response);
      return {
        concepts: Array.isArray(parsed.concepts) ? parsed.concepts : [],
        relationships: Array.isArray(parsed.relationships) ? parsed.relationships : []
      };
    } catch (err) {
      console.error(`[OllamaProvider] Graph extraction failed: ${err.message}`);
      return { concepts: [], relationships: [] };
    }
  }

  getConfig() {
    return {
      type: 'ollama',
      baseUrl: this.baseUrl,
      embeddingModel: this.embeddingModel,
      chatModel: this.chatModel,
      dimensions: this.dimensions
    };
  }
}

File: src/adaptive_memory_engine/providers/ollama_provider.py
"""Ollama provider (local / private)."""

from __future__ import annotations

import json
import logging
import re
from collections.abc import Iterable

import httpx

from .base import IntelligentProvider
from .http import request_with_retry

log = logging.getLogger(__name__)


class OllamaProvider(IntelligentProvider):
    name = "ollama"
    dimensions = 768

    def __init__(
        self,
        host: str = "http://localhost:11434",
        embedding_model: str = "nomic-embed-text",
        chat_model: str = "llama3.2",
        timeout: float = 120.0,
    ) -> None:
        self.host = host.rstrip("/")
        self.embedding_model = embedding_model
        self.chat_model = chat_model
        self._client = httpx.Client(timeout=timeout)

    def is_available(self) -> bool:
        try:
            r = request_with_retry(
                self._client, "GET", f"{self.host}/api/tags", timeout=5.0, max_attempts=1
            )
            return r.status_code == 200
        except Exception:  # noqa: BLE001
            return False

    def embed(self, text: str) -> list[float]:
        r = request_with_retry(
            self._client,
            "POST",
            f"{self.host}/api/embeddings",
            json={"model": self.embedding_model, "prompt": text[:8000]},
        )
        r.raise_for_status()
        return r.json()["embedding"]

    def embed_batch(self, texts: Iterable[str]) -> list[list[float]]:
        # Ollama has no native batch endpoint; sequential
        return [self.embed(t) for t in texts]

    def auto_tag(self, key: str, content: str) -> list[str]:
        prompt = (
            "Generate 3-5 concise, lowercase tags as a JSON array of strings.\n\n"
            f"Key: {key}\nContent: {content[:1500]}\nTags:"
        )
        text = self._generate(prompt, fmt="json")
        return self._parse_tag_list(text)

    def synthesize(self, content: str, task: str, style: str = "concise") -> str:
        prompt = f"Task: {task}\nStyle: {style}\n\nContent:\n{content[:6000]}\n\nResponse:"
        return self._generate(prompt)

    def expand_query(self, query: str) -> list[str]:
        prompt = (
            f"Generate 3 paraphrases as a JSON array of strings.\n\nQuery: {query}\nParaphrases:"
        )
        text = self._generate(prompt, fmt="json")
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("queries") or arr.get("paraphrases") or []
            if isinstance(arr, list):
                return [str(x) for x in arr]
        except Exception:  # noqa: BLE001
            pass
        return [query]

    def get_config(self) -> dict:
        return {
            "type": "ollama",
            "host": self.host,
            "embeddingModel": self.embedding_model,
            "chatModel": self.chat_model,
        }

    def _generate(self, prompt: str, fmt: str | None = None) -> str:
        body: dict = {"model": self.chat_model, "prompt": prompt, "stream": False}
        if fmt:
            body["format"] = fmt
        r = request_with_retry(self._client, "POST", f"{self.host}/api/generate", json=body)
        r.raise_for_status()
        return r.json()["response"]

    def close(self) -> None:
        self._client.close()

    def _parse_tag_list(self, text: str) -> list[str]:
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("tags", [])
            if isinstance(arr, list):
                return [str(t).strip().lower() for t in arr if str(t).strip()][:10]
        except Exception:  # noqa: BLE001
            pass
        return [m.group(1).lower() for m in re.finditer(r'"([^"]+)"', text)][:10]
